content/en/docs/upgrade/devops-data-migration-upgrade.md

```diff
@@ -0,0 +1,104 @@
+---
+title: "KubeSphere DevOps Data Migration during Upgrade"
+keywords: "Kubernetes, Upgrade, KubeSphere, v2.1.0, v3.0.0, DevOps"
+description: "Understand the KubeSphere DevOps data migration and its risks during upgrade."
+linkTitle: "KubeSphere DevOps Data Migration during Upgrade"
+weight: 7700
+---
+
+When upgrading KubeSphere from v2.1 to v3.0, the issues users may encounter in DevOps are mainly caused by the changes in the DevOps architecture. This document illustrates the changes in the DevOps architecture and explores the data migration plans and risks evaluation.
+
+## DevOps Architecture Comparison
+
+### DevOps in KubeSphere v2.1
+
+![devops-2-1-architecture](/images/docs/upgrade/devops-data-migration-upgrade/devops-2-1-architecture.png)
+
+In KubeSphere v2.1, the DevOps data are stored in Jenkins, while the MySQL in KubeSphere stores the authorization information and the corresponding relations among projects, pipelines, and personnel. Frontend users make calls to the Api Gateway to get authentication from the ks-apiserver and then connect to Jenkins.
+
+Another special type is `/kapis/jenkins.kubesphere.io`, which directly passes to Jenkins through the Api Gateway and mainly used when downloading the archive files.
+
+### DevOps in KubeSphere v3.0
+
+![devops-3-0-architecture](/images/docs/upgrade/devops-data-migration-upgrade/devops-3-0-architecture.png)
+
+In KubeSphere v3.0, the DevOps data are stored in CustomResourceDefinition (CRD). The data in the CRD are synchronized to Jenkins in an one-way direction, and any previous data with the same name will be overwritten. There are mainly two types of operation. One type is the operation of creation, and the other is the operation of triggering.
+
+The operation of creation mainly involves three CRD types, including DevOps projects, pipelines and credentials. Users make calls to the interface of ks-apiserver through frontend to create resources and store them in the etcd, and then the ks-controller-manager keeps synchronizing those objects to Jenkins. The operation of triggering mainly involves instant actions, including running a pipeline and reviewing a pipeline. Users make calls to the ks-apiserver through frontend, and then, after data conversion, directly make calls to the Jenkins API.
```

Comment:
->
Creation mainly involves three CRD types, including DevOps projects, pipelines and credentials. Users make calls at the frontend to the interface of ks-apiserver to create resources and store them in etcd, and then ks-controller-manager keeps synchronizing these objects to Jenkins. Triggering mainly entails instant actions, including running a pipeline and reviewing a pipeline. Users make calls at the frontend to ks-apiserver, and then, after data conversion, directly make calls to the Jenkins API.